CONTRACTS: Several fixes for loop contracts

SaswatPadhi opened the pull request:
The main change in this PR is that loop contract instrumentation is now disabled along program paths that result in vacuous loop (i.e., the loop guard doesn't hold initially and the loop never runs).

At a high level, this is equivalent to instrumenting a:

```
while (G) do S;
```

loop as:

```c
if (G) {
  while (G) do S;
}
```

instead. The instrumented code would not be triggered if `G` was initially `false` and we never enter the loop.

This allows for significantly simpler loop contracts. A simple example is attached as a regression test in this PR.

Things get a little tricky when `G` can have side effects, and the instrumented code now jumps back and forth to avoid replicating the entire `G`. I have added several comments to clarify the instrumentation process. We check all side effects in loop guards as well against the loop assigns clauses (see attached regression tests).

Finally, this also allows us to move the instrumentation for decreases clauses to before the loop head (i.e., out of the loop "body"), so we no longer require the `source_location` hack to figure out where the loop guard ends and body starts.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi committed:
CONTRACTS: Several fixes for loop contracts

+ Disabled loop contract instrumentation along program paths that result in vacuous loops,
  i.e., when the loop guard doesn't hold initially and the loop never runs.
  This allows for significantly simpler loop contracts in some cases.
+ Fixed the instrumentation for decreases clauses.
  It no longer requires the `source_location` hack to find the beginning of loop "body".
+ Added some regression tests for loops that have side effects within guards

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -221,8 +202,41 @@ void code_contractst::check_apply_loop_contracts(
+  //
+  //   HEAD:
+  //     if (guard) goto EXIT;
+  //     loop_body;
```
could we make it explicit that the exit GOTO is not necessarily the first instruction of the instructions forming the loop ? (in case the guard is complex or has side effects)

```
HEAD:
eval_guard;
if(!guard) GOTO EXIT
loop_body;
GOTO HEAD;
EXIT:
  ...
```

SaswatPadhi replied:
Yes, good point. I will update it.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -221,8 +202,41 @@ void code_contractst::check_apply_loop_contracts(
+  //     old_decreases_value = decreases_clause_expr;
+  //   HEAD:
+  //     if (guard) goto EXIT;
+  //     loop_body;
```
It seems like in this formulation the loop body in the base case would be checked against snapshots that have been skipped over (as discussed this morning)? Maybe the comment is out of sync with the implementation?
```

SaswatPadhi replied:
Indeed it is out of sync with the implementation. The implementation snapshots things correctly.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -221,8 +202,41 @@ void code_contractst::check_apply_loop_contracts(
+  //     assume (invariant_expr);
+  //     old_decreases_value = decreases_clause_expr;
+  //   HEAD:
+  //     if (guard) goto EXIT;
```
`if (!guard)` ?

SaswatPadhi replied:
Yes 😅

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -232,24 +246,44 @@ void code_contractst::check_apply_loop_contracts(
-  // We use a block scope to create a temporary assertion,
-  // and immediately convert it to goto instructions.
+  // Create a snapshot of the invariant so that we can check the base case,
+  // if the loop is not vacuous must be abstracted with contracts.
```
*and must be abstracted

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -187,26 +187,7 @@ void code_contractst::check_apply_loop_contracts(
 
```
Random place for comment:

At line 150-154 is this method to detect the loop end really reliable reliable ? It seems that the result returned by `natural_loops` already contains for each loop the list of instructions forming the loop.
```
  goto_programt::targett loop_end = loop_head;
  for(const auto &t : loop)
    if(
      t->is_goto() && t->get_target() == loop_head &&
      t->location_number > loop_end->location_number)
      loop_end = t;
```

SaswatPadhi replied:
Yes it captures a "set" of all the instructions (not "list"), but also, we need to find the "loop_end" instruction where we have a back edge (so that we can remove it).

As to whether this loop is reliable for finding the "loop_end", I think it is? Do you see a case where this may not work?

remi-delmas-3000 replied:
I was thinking we should maybe search for loop end within the set of instructions returned by natural loops. In the general case don't think that loops always correspond to spans of consecutive instructions.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -187,26 +187,7 @@ void code_contractst::check_apply_loop_contracts(
 
```
Random place for comment : at Line 170 and 179 what about default values for other clauses (assigns, decreases)

SaswatPadhi replied:
Assigns clause is always optional (since we try to infer it), so no message is shown regarding that. I added a message regarding missing decreases clause -- that termination will not be verified.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -232,24 +246,44 @@ void code_contractst::check_apply_loop_contracts(
+      loop_head->source_location();
+    converter.goto_convert(
+      initial_invariant_value_assignment, generated_code, mode);
   }
```
should we check here if the invariant contains `side_effect_exprt`  or `side_effect_function_callt` nodes ?

SaswatPadhi replied:
The goto compiler (`goto-cc`) guards against it, but I guess yeah we could double check here just to guard against some other instrumentation step introducing invariants with side effects.

To be consistent, I think we should guard it the same way as the compiler does? `!has_subexpr(invariant, ID_side_effect)`?

SaswatPadhi replied:
Per our offline discussion, setting this aside for a separate PR. We need to handle this consistently for all function calls within all contracts and in a separate PR we will handle calls (to pure + total functions) within function & loop contracts together.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -221,8 +202,41 @@ void code_contractst::check_apply_loop_contracts(
     return invariant_copy;
   };
```
Why do we need a closure ? Does this transformation need to be applied before or after the history variable replacement inside `invariant` ?

SaswatPadhi replied:
The closure if for processing quantifiers correctly. Every time we use the invariant expression, we have to use a “fresh” expression because otherwise the temporary variables introduced for quantifiers will clash.

Consider `__CPROVER_loop_invariant ( __CPROVER_forall { int i ; … })`. When we process it, we don’t actually have a special scope for `i` in the goto program. We introduce a new temporary in the current scope and rewrite (replace `i`) in the invariant. If we don’t create a “fresh” `i` for each occurrence of the invariant expression (2 assertions, 1 assumption) then, for example, the temporary for `i` from base case assertion will clash with the one in the step case assertion

tautschnig replied:
No, you don't have to worry about the uniqueness of `i` within quantifiers: the back-end takes care of this, i.e., the scope of an identifier introduced within a quantifier is just that quantified expression. This was fixed in d9b6f156f69e.

SaswatPadhi replied:
No, I believe this is different. I tried disabling this and using `invariant` expression directly everywhere and it fails exactly as it did before. This is my understanding:

For quantifiers that appear within `assert` and `assume` statements, the goto-cc compiler introduces new temporary variables during compilation. However, for quantifiers that appear within function and loop contracts, since the are never compiled, not until instrumentation, temporaries are not introduced during compilation. At the instrumentation step, we introduce a unique temporary for each unique quantified variable, in each quantified expression. I could just make 3 copies of the invariant, each with its own quantified variables, but I used this lamda to reduce code duplication and construct them on demand.

SaswatPadhi replied:
@tautschnig Thanks! Yes indeed this works as expected now. Just introducing a single temporary for each unique quantified variable works fine now, even when thise quantified expression is used more than once.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,9 +345,25 @@ void code_contractst::check_apply_loop_contracts(
-    goto_function.body, loop_head, snapshot_instructions);
+  // FIXME: this is not true for write set targets that
+  // might depend on other write set targets.
+  generated_code.destructive_append(snapshot_instructions);
```
It seems like  `to_havoc` and `instrument_spec_assigns` almost contain the same thing, except for local statics detected at line 294 which are not added to `to_havoc`

SaswatPadhi replied:
This is fixed now.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -331,23 +381,11 @@ void code_contractst::check_apply_loop_contracts(
+  goto_function.body.destructive_insert(loop_head, generated_code);
 
   // Generate: assume(invariant) just after havocing
   // We use a block scope to create a temporary assumption,
```
at line 394 it is the second time this closure `invariant_expr()` gets called, is there a risk in executing the side effects of this closure twice ? or is it the very reason this is done using a closure ? Why not transform the invariant expression once and for all in the beginning ?

SaswatPadhi replied:
resolved now; I was not aware of the fix that @tautschnig pointed out

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -331,23 +381,11 @@ void code_contractst::check_apply_loop_contracts(
+  // We do not `add_pragma_disable_assigns_check`,
+  // so that the enclosing scope's assigns clause instrumentation
+  // would pick these up for inclusion (subset) checks.
+  goto_function.body.destructive_insert(loop_head, generated_code);
```
could we use separate variables for the different insertion steps ? This would make it clearer what gets inserted where.

SaswatPadhi replied:
Yeah that's a good point. I will make this change. It was simple in the beginning, but now that we have 3 different contract types it's quite messy :D

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
+  // Insert all the generated code so far:
+  // Assume invariant, decl the variant temporaries,
+  // and initialize old variant values.
+  goto_function.body.destructive_insert(
```
Why do we use three distinct destructive inserts instead of one ? it seems like the results would be exacly the same

SaswatPadhi replied:
This is because a block of code in the middle (the havocing code) shouldn't be pragma-fied. The block above and below it should be (and their order matters), to disable assigns clause checking of the enclosing block on them. So right now the block is split into 3 blocks with distinct pragma annotations and we call insert thrice.
I added lots of comments now and this would hopefully be clear in the code now.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
+  // Generate: assert(invariant) just before the iteration ends
   // We use a block scope to create a temporary assertion,
   // and immediately convert it to goto instructions.
   {
```
third time the closure invariant_expr() is invoked, could we document why this is used and what it does ?

SaswatPadhi replied:
resolved now; I was not aware of the fix that @tautschnig pointed out

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
+  // because of the unconditional jump above.
+
+  // Generate: assert(invariant) just before the iteration ends
   // We use a block scope to create a temporary assertion,
```
What does this mean ? why is the assertion called "temporary" ?

SaswatPadhi replied:
So as not to pollute the function's namespace. The assertion is not temporary. It is immediately ~~destroyed when we~~ `goto_convert`ed, and there's no point keeping the `codet` object around.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
+  // The following code is only reachable in the step case,
+  // i.e., when in_base_case == false,
+  // because of the unconditional jump above.
+
```
There are a few variables like `__init_invariant` or `__in_base_case` that are never declared DEAD

SaswatPadhi replied:
Good point; will fix

SaswatPadhi replied:
Hmm, I tried doing this, but it seems a little tricky.

These variables should go dead when we exit the loop, either by falsifying the guard, or after `break`s, or after `goto`s. Should we add these dead statements at each of these exit targets?

remi-delmas-3000 replied:
so at each jump that does not go to an instruction present in the `loop` set, or when the next instruction is not in the `loop` set ?

remi-delmas-3000 replied:
Was this eventually solved ?

SaswatPadhi replied:
Yes, this is fixed now.

remi-delmas-3000 reviewed (changes requested):
Declare all instrumentation DEAD after the loop, add doc on the invariant_expr() closure, use distinct variables for `generated_code` before and after loop to make it easier to understand what is going on. Keep comments in sync with the implementation.

remi-delmas-3000 commented:
Other comment: the havoc statements do not contain the local static variables

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +358,154 @@ void code_contractst::check_apply_loop_contracts(
-    goto_function.body, loop_head, snapshot_instructions);
-
-  optionalt<cfg_infot> cfg_empty_info;
+  // FIXME: this is not true for write set targets that
+  // might depend on other write set targets.
```
Is there a way to check if we are in this case with the specified or inferred targets ? 
And is widening now applied right after the targets have been inferred or still during havocing ?

SaswatPadhi replied:
The widening part is not yet fixed.

I could add a separate commit to fix this PR, but is this still high priority given that we double check all inferred targets anyway by instrumenting the loop?

remi-delmas-3000 replied:
Right now loop contracts still use the `havoc_assigns_targetst` class because it contains the widening logic.
 
Widening right after the inference would allow to use the new class `havoc_assigns_clause_targetst` which shares the CAR snapshotting logic with assigns clause checking.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +358,154 @@ void code_contractst::check_apply_loop_contracts(
-    generated_code.add(goto_programt::make_goto(
-      loop_end,
-      side_effect_expr_nondett(bool_typet(), loop_head->source_location())));
+    pre_loop_head_instrs.add(goto_programt::make_goto(
+      loop_end, side_effect_expr_nondett(bool_typet(), loop_head_location)));
```
What does this jump model exactly ? Is this the sound thing to do ?

SaswatPadhi replied:
It's right after a `throw 0`, so ignored for now. This needs to be fixed for handling `do ... while` loop contracts, but the semantics of loop invariants on those loops is still under discussion (whether the firs iteration should be "pre" code or "loop" code).

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -450,44 +514,60 @@ void code_contractst::check_apply_loop_contracts(
+  // TODO: ^^^ I think these should be inserted at all exit points:
+  // falsification of guard, break statements, or even `goto` jumps
+  // out of a loop.
+  // Is there an easy way to add these everywhere?
```
at any point where the jump target or the next instruction is not in the `loop` set ?

SaswatPadhi replied:
Thanks. Implemented it as you suggested.

SaswatPadhi committed:
CONTRACTS: Lots of sanity checks on loop contracts

Our loop contracts instrumentation routine makes some assumptions regarding the structure of loops.
In this commit, I have added checks to verify them before we start instrumenting,
so as to avoid potentially unsound instrumentation.

feliperodri commented on regression/contracts/loop_contracts_reject_overlapping_loops/test.desc:
```diff
@@ -0,0 +1,11 @@
+main.c
+--apply-loop-contracts
+activate-multi-line-match
+Inner loop at: file main.c line 12 function main does not have contracts, but an enclosing loop does.\nPlease provide contracts for this loop, or unwind it first.
```
Why `\n`? I guess this is a typo from the error string.

SaswatPadhi replied:
No `\n` is for multi-line matching.
The error message is split over two lines and ` activate-multi-line-match` requires the newline to be explicit.

feliperodri commented on regression/contracts/nonvacuous_loop_contracts/main.c:
```diff
@@ -0,0 +1,36 @@
+int main()
+{
+  size_t size;
+  __CPROVER_assume(size <= MAX_SIZE);
```
nitpick: would it be nice a comment here.

SaswatPadhi replied:
I realized that we don't need this `MAX_SIZE` at all, removing it.

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.h:
```diff
@@ -451,6 +451,17 @@ class instrument_spec_assignst
     skip_function_paramst skip_function_params,
     optionalt<cfg_infot> &cfg_info_opt);
 
+  template <typename C>
+  void get_static_locals(std::insert_iterator<C> inserter) const
```
Where is the documentation header for this method?

SaswatPadhi replied:
Good point, I'll add doc

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -140,123 +140,140 @@ void code_contractst::check_apply_loop_contracts(
+  // loop assigns check     |      initial_invariant_val = invariant_expr;
+  //  - unchecked, temp     |      in_base_case = true;
+  // func assigns check     |      snapshot (write_set);
+  //  - disabled via pragma |      goto HEAD;
```
disable what via pragma?

remi-delmas-3000 replied:
`func assigns check` is disabled by pragma

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -140,123 +140,140 @@ void code_contractst::check_apply_loop_contracts(
-  // Generate: assert(invariant) just before the loop
-  // We use a block scope to create a temporary assertion,
-  // and immediately convert it to goto instructions.
+  // Create a temporary to track if we enter a loop, i.e., get pas its guard.
```
get past its guard.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -140,123 +140,140 @@ void code_contractst::check_apply_loop_contracts(
+    // Although the invariant at this point will not have side effects,
+    // it is still a C expression, and needs to be "goto_convert"ed.
+    // Note that this conversion may emit many GOTO instructions,
+    // and we cannot simply do:
+    //    pre_loop_head_instrs.add(goto_programt::make_assignment(...))
```
why?

remi-delmas-3000 replied:
We should just remove the part "and we cannot simply do ...". The reason is already given above (the invariant expression is still a C expression and needs to be converted to a basic GOTO expression)

SaswatPadhi replied:
Okay I can remove it.

> why?

Because `make_assignment` generates a single GOTO instruction and doesn't actually goto_convert the subexpressions. So in this case the assignment instruction could be potentially illformed.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +330,157 @@ void code_contractst::check_apply_loop_contracts(
   // Generate: assume(invariant) just after havocing
-  // We use a block scope to create a temporary assumption,
-  // and immediately convert it to goto instructions.
+  // We use a block scope for assumption, since it's immediately goto converted,
+  // and doesn't need to be kept around.
```
Nice

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +330,157 @@ void code_contractst::check_apply_loop_contracts(
     new_decreases_vars.push_back(new_decreases_var);
   }
 
   // TODO: Fix loop contract handling for do/while loops.
```
Do we track this somewhere? We are using "FIXME" and "TODO" as labels for such comments. I'd pick one and make consistent across this file.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +330,157 @@ void code_contractst::check_apply_loop_contracts(
-    generated_code.add(goto_programt::make_goto(
-      loop_end,
-      side_effect_expr_nondett(bool_typet(), loop_head->source_location())));
+    // pre_loop_head_instrs.add(goto_programt::make_goto(
+    //   loop_end, side_effect_expr_nondett(bool_typet(), loop_head_location)));
```
Why keep this around?

SaswatPadhi replied:
When we fix the `do ... while` case, we might actually want to do something similar, so I just kept it around

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -311,137 +330,157 @@ void code_contractst::check_apply_loop_contracts(
-
   // Generate: assignments to store the multidimensional decreases clause's
-  // value just before the loop body (but just after the loop guard)
+  // value just before the loop head
```
`loop_head`

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -450,44 +489,82 @@ void code_contractst::check_apply_loop_contracts(
   loop_end->set_condition(boolean_negate(loop_end->get_condition()));
+
+  std::set<goto_programt::targett> seen_targets;
+  // Find all exit points of the loop, make temporary variables `DEAD`,
+  // and check that step case was checked for non-vacuous loops.
+  for(const auto &t : loop)
+  {
+    if(!t->is_goto())
+      continue;
+
+    auto exit_target = t->get_target();
+    if(
+      loop.contains(exit_target) ||
+      seen_targets.find(exit_target) != seen_targets.end())
+      continue;
+
+    seen_targets.insert(exit_target);
+
+    goto_programt pre_loop_exit_instrs;
+    // Assertion to check that step case was checked if we entered the loop.
+    pre_loop_exit_instrs.add(goto_programt::make_assertion(
+      or_exprt{not_exprt{entered_loop}, not_exprt{in_base_case}},
+      loop_head_location));
+    pre_loop_exit_instrs.instructions.back()
+      .source_location_nonconst()
+      .set_comment("Check that loop instrumentation was not truncated");
+    // Instructions to make all the temporaries go dead.
+    pre_loop_exit_instrs.add(
+      goto_programt::make_dead(in_base_case, loop_head_location));
+    pre_loop_exit_instrs.add(
+      goto_programt::make_dead(initial_invariant_val, loop_head_location));
+    for(const auto &v : history_var_map)
+    {
+      pre_loop_exit_instrs.add(
+        goto_programt::make_dead(to_symbol_expr(v.second), loop_head_location));
+    }
+    // Insert these instructions, preserving the loop end target.
+    insert_before_swap_and_advance(
+      goto_function.body, exit_target, pre_loop_exit_instrs);
+  }
```
nitpick: this code could have gone into a separate function. In fact, this method has over 400 LOC and could be break down into smaller modules that would represent the major steps of our instrumentation.

remi-delmas-3000 replied:
+1

SaswatPadhi replied:
I agree. My plan is to refactor this out into `instrument_loop_contracts.cpp` in the next PR

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+        loop_end = t;
+    }
+
+    if(loop_end == loop_head)
+    {
+      log.error() << "Could not find end of the loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw 0;
+    }
```
Case not possible to be covered via regression tests?

SaswatPadhi replied:
No, this is more of a consistency check. The only way this would be triggered is if we compute natural loops, then instrument the loops and somehow remove the back edges, and then we try computing the loop_end

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+      invariant = conjunction(invariant.operands());
+      if(decreases_clause.is_nil())
+      {
+        log.warning() << "The loop at "
+                      << loop_head->source_location().as_string()
+                      << " does not have a decreases clause in its contract.\n"
+                      << "Termination of this loop will not be verified."
+                      << messaget::eom;
```
Nice

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+      if(i->is_goto() && !loop_content.contains(i->get_target()))
+        continue;
+
+      // `assume false`, is okay
+      if(i->is_assume() && i->get_condition().is_false())
+        continue;
+
```
Not covered by regression tests?

SaswatPadhi replied:
On my machine I see that we hit this line on regression test `github_6168_infinite_unwinding_bug`.

If I disable these 2 statements that regression test crashes. I think this might be a bug with coverage CI job

remi-delmas-3000 replied:
`assume(false)` are usually the result of automatic function pointer replacement.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+        continue;
+      }
+
+      log.error() << "Cannot enforce contracts on irregular loop at: "
+                  << loop_head->source_location()
+                  << ".\nLoop body must be a contiguous set of instructions."
+                  << messaget::eom;
+      log.debug() << "Out-of-loop-body instruction:" << messaget::eom;
+      goto_function.body.output_instruction(ns, function_name, log.debug(), *i);
+      throw 0;
+    }
```
This is not covered by regression tests because is design to catch inconsistent GOTO programs? I'd add a comment if that's true.

SaswatPadhi replied:
Yes exactly. I tried to, but couldn't write loops in C that would trigger this.

I will add a comment.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -920,23 +1121,69 @@ void code_contractst::apply_loop_contract(
+      if(loop_nesting_graph[outer].content.contains(
+           loop_nesting_graph[inner].head_target))
+      {
+        if(!loop_nesting_graph[outer].content.contains(
+             loop_nesting_graph[inner].end_target))
+        {
+          log.error()
+            << "Overlapping loops at:\n"
+            << loop_nesting_graph[outer].head_target->source_location()
+            << "\nand\n"
+            << loop_nesting_graph[inner].head_target->source_location()
+            << "\nLoops must be nested or sequential for contracts to be "
+               "enforced."
+            << messaget::eom;
+        }
```
Why not covered by regression tests?

SaswatPadhi replied:
I don't think we can write loops in C which would trigger this check. It can only be triggered if someone messes with the goto instructions directly and introduces some statements in the middle of a loop.

We should have this check in place though because our instrumentation code assumes this property and would be unsound if two loop bodies somehow overlap.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -920,23 +1121,69 @@ void code_contractst::apply_loop_contract(
     }
   }
 
+  // make sure all children of a contractified loop also have contracts
```
nitpick: `Make sure all children of a contractified loop also have contracts.`

feliperodri reviewed (approved):
LGTM, no blocking comments.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -450,44 +489,82 @@ void code_contractst::check_apply_loop_contracts(
+  // and check that step case was checked for non-vacuous loops.
+  for(const auto &t : loop)
+  {
+    if(!t->is_goto())
```
Are all loop exit points necessarily GOTOs instructions?

Isn't it possible to exit the loop just by executing the instruction that comes right after an instruction that is in the loop in sequential order ?

remi-delmas-3000 replied:
Could we try the condition 
"the instruction is a GOTO with a target that is not in the set of loop instructions OR the next instruction in sequential order is not in the set of loop instructions ?"

SaswatPadhi replied:
> Are all loop exit points necessarily GOTOs instructions?

I believe so.

> Isn't it possible to exit the loop just by executing the instruction that comes right after an instruction that is in the loop in sequential order ?

Just because we have some instructions after the loop, doesn't mean they would be executed automatically, because the last instruction of a loop is a back edge GOTO, and the instructions that follow might be unreachable. We must either jump out via a `break` or `return` or a `goto` on loop guard falsification at the loop head -- all 3 of these emit `GOTO` instructions.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -450,44 +489,82 @@ void code_contractst::check_apply_loop_contracts(
+      pre_loop_exit_instrs.add(
+        goto_programt::make_dead(to_symbol_expr(v.second), loop_head_location));
+    }
+    // Insert these instructions, preserving the loop end target.
```
Double checking: does this guarantee that a jump that was previously to `exit_target` will jump into the freshly inserted DEAD instructions instead ?

SaswatPadhi replied:
Yes

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+
+    to_check_contracts_on_children.push_back(idx);
+
+    // If the loop has contracts to be enforced, then
```
Proposal:

"Check that all instructions found in sequential order between `loop_head` and  `loop_end` belong to the set of instructions computed by the `natural_loop` analysis.

We also allow the presence of:
- `assume(false)` instructions which are introduced by function pointer or nested loop transformations, and have no successor instructions
- `SET_RETURN_VALUE` instructions followed by an uninterrupted sequence of  `DEAD` instructions and a GOTO instruction that jumps out of the loop, which model `return` statements in GOTO.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+    to_check_contracts_on_children.push_back(idx);
+
+    // If the loop has contracts to be enforced, then
+    // reject irregular loops that have "holes" within their body,
```
We don't really have a  notion of  'irregular' loop.  Could we rephrase the comment and move this check in a separate method ?

We could add a doc like this for the method :
```
By definition the loop body is the set of instructions computed by `natural_loops` based on the CFG, and here we check that straight line code or instructions from different loops do not mix and overlap in the serialized representation.
This check is needed because we perform the instrumentation by traversing the sequence of instructions and we want to make sure we are only going to instrument loop instructions or function exit instructions.

So we check that all instructions found in sequential order between loop_head and loop_end belong to the set of instructions computed by the natural_loop analysis, while allowing the presence of:
- `assume(false)` instructions which are introduced by function pointer or nested loop transformations, and have no successor instructions
- `SET_RETURN_VALUE` instructions followed by an uninterrupted sequence of `DEAD` instructions and a `GOTO` instruction that jumps out of the loop, which model return statements in GOTO.
```

If in addition we could check that instructions are in topological order this would prove that this is equivalent to traversing the CFG to do the instrumentation.

remi-delmas-3000 commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -920,23 +1121,69 @@ void code_contractst::apply_loop_contract(
   for(const auto &idx : loop_nesting_graph.topsort())
   {
     const auto &loop_node = loop_nesting_graph[idx];
+    if(
```
Considering the check performed at line 1151 is this case really possible ?

Does this correspond to a case where we have a loop without contract ? What happens if this loop is itself embedded in an outer loop which has a contract ? 

Would this result in applying the loop contract on the outer loop but without ensuring that the body of the outer loop is loop-free ?

SaswatPadhi replied:
> Considering the check performed at line 1151 is this case really possible ?

Yes. Line 1151 checks that all loops within an outer loop with contracts also have contracts. We throw an error and stop instrumentation if we find a single violation.

This line skips instrumentation for the (outer) loops that don't have contracts.

> Does this correspond to a case where we have a loop without contract ? What happens if this loop is itself embedded in an outer loop which has a contract ?

1. Yes
2. It would have been checked at line 1151

> Would this result in applying the loop contract on the outer loop but without ensuring that the body of the outer loop is loop-free ?

I'm not sure how that would be possible.

remi-delmas-3000 reviewed (changes requested):
move the loop dependency graph construction and loop instruction checking to a dedicated method with documentation, and add clarification on the case when a loop does not have a contract and we skip instrumentation (L1167)
